codex/blockexchange/engine/engine.nim

```diff
@@ -290,6 +290,8 @@ proc resolveBlocks*(b: BlockExcEngine, blocksDelivery: seq[BlockDelivery]) {.asy
     cids.incl(bd.blk.cid)
     if bd.address.leaf:
       cids.incl(bd.address.treeCid)
+
+  asyncSpawn b.issueCancellations(blocksDelivery.mapIt(it.address))
```

Comment:
`asyncSpawn` isn't needed here at all... what I would do is just a regular `await b.issueCancellations(blocksDelivery.mapIt(it.address))`, but issue the network requests in "parallel", collect the futures into a list and then await them all.

Something like this:

```nim
let
  addrs = blocksDelivery.mapIt(it.address)
  failed = (await allFinished(b.peers.mapIt(
  b.network.request.sendWantList(id = it.id,
      addresses = addrs,
      cancel = true) )))
      .filterIt( it.failed )

if failed.len > 0:
  ...
```

`asynSpawn` is used for "unattended" tasks, such as running something periodically, for the duration of the process, etc... It also has a few gotchas, mainly that, if all exceptions aren't handled in the async task runner, then it will crash chronos, because there is no `await` point to propagate the exception to. 

I would change this to something:

Reply:
Ah yeah good point. I reasoned that I didn't want to "block" whatever came next in the delivery flow as cancellation doesn't have an ordering constraint with respect to anything else, really, but the exception handling is a problem. 

On your second point though: do we really want to issue multiple calls to sendWantList? I don't see how this helps: batching the send as a single call should be just as fast, probably lower overhead on the client (fewer messages), and the code is shorter?

Reply:
Btw, I'd probably not do much in terms of error handling here than logging the error. This is a pure-push protocol, and guarantees are probably limited to what the underlying library can provide (i.e., I'm guessing libp2p probably has some retransmission policy if things fail, and that's probably the best we can do here). A full treatment of this would involve reconciling knowledge on first connection, and expiring very old requests.

Reply:
> do we really want to issue multiple calls to sendWantList?

No we definitely don't that and it isn't whats going on either. What my example does is, initiate transmission without waiting for the previous request to complete and it's doing so only once for each peer.

> I reasoned that I didn't want to "block" whatever came next in the delivery flow as cancellation doesn't have an ordering constraint with respect to anything else, really, but the exception handling is a problem.

Yes,  precisely, the snippet above accomplishes what you were trying to do with the `asyncSpawn`, but in a more reliable and less error prone way. 

Really, if you think about it, the `asyncSpawn` + `await` kinda does that as well, but in a bit of a more convoluted way, that `await` in the for loop, will simply serialize the sends to each peer, which isn't required in this case.

```nim
  for neighbor in b.peers:
    await b.network.request.sendWantList(
      id = neighbor.id,
      addresses = addrs,
      cancel = true
    )
```

> Btw, I'd probably not do much in terms of error handling here other than logging the error.

Yeah, but we could probably either disconnect the failed peers or simply try to retransmit it by pushing the same task on the queue.

> I'm guessing libp2p probably has some retransmission policy if things fail,

No, not at all and frankly it's a good thing.

Reply:
> No we definitely don't that and it isn't whats going on either. What my example does is, initiate transmission without waiting for the previous request to complete and it's doing so only once for each peer.
> ...
> Really, if you think about it, the `asyncSpawn` + `await` kinda does that as well, but in a bit of a more convoluted way, that `await` in the for loop, will simply serialize the sends to each peer, which isn't required in this case.

Oh, right, I managed to completely misread the snippet you put there. :-) Yes, that's indeed much better, thank you.

> > Btw, I'd probably not do much in terms of error handling here other than logging the error.
> 
> Yeah, but we could probably either disconnect the failed peers or simply try to retransmit it by pushing the same task on the queue.

Hm... I think outright disconnecting because one single message didn't go through might be too aggressive, isn't it? OTOH retrying means you need a circuit-breaking strategy so you don't end up flooding your own queue with retries for a dead peer (which means tracking retry counters/timers, putting together retry logic, etc).

> > I'm guessing libp2p probably has some retransmission policy if things fail,
> 
> No, not at all and frankly it's a good thing.

Yeah, I'm a bit torn on this. OTOH retry magic may create surprising behavior, on the other having to implement possibly similar retry strategies for every send is also a huge burden.

Reply:
> having to implement possibly similar retry strategies for every send is also a huge burden.

Sure, but that doesn't mean it belongs in the core p2p stack and the biggest challenge is to make it work well with all the backpressure mechanisms that libp2p relies on. In any case, this is not where I would put retries unless there are really strong arguments for it.